src/core/struct_tree.js

```diff
@@ -67,6 +69,439 @@ class StructTreeRoot {
+    newRefs.push({ ref: kidsRef, data: buffer.join("") });
+
+    if (numsRef) {
+      buffer.length = 0;
+      await writeObject(numsRef, nums, buffer, xref);
+      newRefs.push({ ref: numsRef, data: buffer.join("") });
+    }
+
+    buffer.length = 0;
+    await writeObject(parentTreeRef, parentTree, buffer, xref);
+    newRefs.push({ ref: parentTreeRef, data: buffer.join("") });
+
+    buffer.length = 0;
+    await writeObject(structTreeRootRef, structTreeRoot, buffer, xref);
+    newRefs.push({ ref: structTreeRootRef, data: buffer.join("") });
+  }
+
+  static async #writeKids(
+    newAnnotationsByPage,
+    structTreeRootRef,
+    kids,
+    nums,
+    xref,
+    catalog,
+    newRefs,
+    buffer
+  ) {
+    const objr = Name.get("OBJR");
+    let nextKey = -Infinity;
+
+    for (const [pageIndex, elements] of newAnnotationsByPage) {
+      const [, pageRef] = await catalog.getPageDict(pageIndex);
+      for (const {
+        accessibilityData: { type, title, lang, alt, expanded, actualText },
+        ref,
+        parentTreeId,
+        structTreeParent,
+      } of elements) {
+        nextKey = Math.max(nextKey, parentTreeId);
+
+        const tagRef = xref.getNewTemporaryRef();
+        const tagDict = new Dict(xref);
+
+        // The structure type is required.
+        tagDict.set("S", Name.get(type));
+
+        if (title) {
+          tagDict.set("T", title);
+        }
+        if (lang) {
+          tagDict.set("Lang", lang);
+        }
+        if (alt) {
+          tagDict.set("Alt", alt);
+        }
+        if (expanded) {
+          tagDict.set("E", expanded);
+        }
+        if (actualText) {
+          tagDict.set("ActualText", actualText);
+        }
+
+        if (structTreeParent) {
+          await this.#updateParentTag(
+            structTreeParent,
+            tagDict,
+            tagRef,
+            xref,
+            newRefs,
+            buffer
+          );
+        } else {
+          tagDict.set("P", structTreeRootRef);
+        }
+
+        const objDict = new Dict(xref);
+        tagDict.set("K", objDict);
+        objDict.set("Type", objr);
+        objDict.set("Pg", pageRef);
+        objDict.set("Obj", ref);
+
+        buffer.length = 0;
+        await writeObject(tagRef, tagDict, buffer, xref);
+        newRefs.push({ ref: tagRef, data: buffer.join("") });
+
+        nums.push(parentTreeId, tagRef);
+        kids.push(tagRef);
+      }
+    }
+    return nextKey + 1;
+  }
+
+  static #collectParents(elements, xref, pageDict, pageIndex, parentTree) {
+    const idToElement = new Map();
+    for (const element of elements) {
+      if (element.structTreeParentId) {
+        const id = parseInt(element.structTreeParentId.split("_mc")[1], 10);
+        idToElement.set(id, element);
+      }
+    }
+
+    const id = pageDict.get("StructParents");
+    const numberTree = new NumberTree(parentTree, xref);
+    const parentArray = numberTree.get(id);
+    if (Array.isArray(parentArray)) {
```

Comment:
Can we please change this to an early return to reduce the overall indentation level below?
```suggestion
    if (!Array.isArray(parentArray)) {
      return;
    }
```